### PR TITLE
Fixed interface to data iterator kwargs generator + typing

### PR DESCRIPTION
**Pull request by jordis-ai2**

### CODE REVIEW

**Comment by jordis-ai2** [comment]
In the former design, the kwargs generator was not aware of the current worker and had therefore limited ability to provide arguments that could not be directly passed in the instantiation of the offpolicy component

File: core/algorithms/onpolicy_sync/engine.py [hunk 1]
```diff
@@ -624,13 +624,18 @@ def __init__(
         self.last_log: Optional[int] = None
         self.last_save: Optional[int] = None
 
-    def advance_seed(self, seed: Optional[int]) -> Optional[int]:
+    def advance_seed(
+        self, seed: Optional[int], call_worker_seeds=True
+    ) -> Optional[int]:
         if seed is None:
             return seed
         seed = (seed ^ (self.training_pipeline.total_steps + 1)) % (
```
Comment (Lucaweihs) [inline core/algorithms/onpolicy_sync/engine.py:632]: Shouldn't this be `(seed ** (self.training_pipeline.total_steps + 1)) % (2 ** 31 - 1)`. The `^` is a bitwise operation and is pretty strange. If so we should have it be `pow(seed, self.training_pipeline.total_steps + 1, 2 ** 31 - 1)` as this will make things more efficient.

File: core/algorithms/onpolicy_sync/engine.py [hunk 2]
```diff
@@ -624,13 +624,18 @@ def __init__(
...
             2 ** 31 - 1
         )  # same seed for all workers
 
+        if not call_worker_seeds:
+            return seed
+
```
Comment (Lucaweihs) [inline core/algorithms/onpolicy_sync/engine.py:638]: I like the idea of having something like this but can we rename `call_worker_seeds` to something like `return_same_seed_per_worker` and then rather than immediately returning seed here we could instead change 
```python
if self.mode == "train" or self.mode == "test":
```
to
```python
if (not return_same_seed_per_worker) and (self.mode == "train" or self.mode == "test"):
```
below? I like this slightly better as things will be a bit more consistent I think.

File: core/algorithms/onpolicy_sync/engine.py [hunk 3]
```diff
@@ -906,13 +911,24 @@ def update(self, rollouts: RolloutStorage):
...
-        ), "current_worker is reserved for the engine to be passed to the offpolicy data iterator builder"
-        return data_iterator_builder(current_worker=self.worker_id, **kwargs)
+
+        if self.num_workers == 1:
+            rollouts_per_worker = [self.num_samplers]
+        else:
+            rollouts_per_worker = self.num_samplers_per_worker
+
+        # common seed for all workers (in case we wish to shuffle the full dataset before iterating on one partition)
+        seed = self.advance_seed(self.seed, call_worker_seeds=False)
```
Comment (Lucaweihs) [inline core/algorithms/onpolicy_sync/engine.py:925]: This is a great idea.

File: scripts/build_docs.py [hunk 4]
```diff
@@ -244,7 +244,7 @@ def recursively_create_nav_structure(nested_dict):
         if d in git_dirs:
             git_dirs.remove(d)
 
-    threads = []
+    threads: List = []
```
Comment (Lucaweihs) [inline scripts/build_docs.py:247]: Thanks, I obviously forgot to run mypy.

File: core/algorithms/onpolicy_sync/engine.py (hunk 1 above)
Comment (jordis-ai2) [reply core/algorithms/onpolicy_sync/engine.py:632]: We've discussed this already a couple of times. I actually designed it as a bitwise operation and checked that the seeds were not too correlated and things looked fine. I'm reluctant to change something that works. Bitwise XORs are pretty common in RNG internals...

File: core/algorithms/onpolicy_sync/engine.py (hunk 3 above)
Comment (jordis-ai2) [reply core/algorithms/onpolicy_sync/engine.py:925]: Yes. It's not straightforward to use (we need to move around the RNG state), but I think it can be useful.

**Comment by jordis-ai2** [comment]
Feel free to merge if the changes seem acceptable!

File: core/algorithms/onpolicy_sync/engine.py (hunk 1 above)
Comment (Lucaweihs) [reply core/algorithms/onpolicy_sync/engine.py:632]: It just surprises me anew every time :) Fair enough, can you add a comment for my sanity saying explicitly that this is meant to be a bitwise operation so I don't make the comment in a few months again?

Comment (Lucaweihs) [reply core/algorithms/onpolicy_sync/engine.py:632]: I'll add this.
